src/Compiler/Microsoft.AspNetCore.Razor.Language/src/Legacy/AutoCompleteEditHandler.cs

```diff
@@ -5,34 +5,38 @@
 
 using System;
 using System.Collections.Generic;
+using System.Diagnostics;
 using Microsoft.AspNetCore.Razor.Language.Syntax;
 using Microsoft.Extensions.Internal;
 
 namespace Microsoft.AspNetCore.Razor.Language.Legacy;
 
 internal class AutoCompleteEditHandler : SpanEditHandler
 {
-    private static readonly int TypeHashCode = typeof(AutoCompleteEditHandler).GetHashCode();
-
-    public AutoCompleteEditHandler(Func<string, IEnumerable<Syntax.InternalSyntax.SyntaxToken>> tokenizer)
-        : base(tokenizer)
+    public static void SetupBuilder(SpanEditHandlerBuilder builder, Func<string, IEnumerable<Syntax.InternalSyntax.SyntaxToken>> tokenizer, bool autoCompleteAtEndOfSpan, out AutoCompleteStringAccessor autoCompleteStringAccessor)
     {
+        var accessor = new AutoCompleteStringAccessor();
+        autoCompleteStringAccessor = accessor;
+        builder.Factory = (acceptedCharacters, tokenizer) => new AutoCompleteEditHandler(accessor)
+        {
+            AcceptedCharacters = acceptedCharacters,
+            Tokenizer = tokenizer,
+            AutoCompleteAtEndOfSpan = autoCompleteAtEndOfSpan,
+        };
     }
 
-    public AutoCompleteEditHandler(Func<string, IEnumerable<Syntax.InternalSyntax.SyntaxToken>> tokenizer, bool autoCompleteAtEndOfSpan)
-        : this(tokenizer)
-    {
-        AutoCompleteAtEndOfSpan = autoCompleteAtEndOfSpan;
-    }
+    private static readonly int TypeHashCode = typeof(AutoCompleteEditHandler).GetHashCode();
+
+    private readonly AutoCompleteStringAccessor _autoCompleteStringAccessor;
 
-    public AutoCompleteEditHandler(Func<string, IEnumerable<Syntax.InternalSyntax.SyntaxToken>> tokenizer, AcceptedCharactersInternal accepted)
-        : base(tokenizer, accepted)
+    private AutoCompleteEditHandler(AutoCompleteStringAccessor autoCompleteStringAccessor)
     {
+        _autoCompleteStringAccessor = autoCompleteStringAccessor;
     }
 
-    public bool AutoCompleteAtEndOfSpan { get; }
+    public required bool AutoCompleteAtEndOfSpan { get; init; }
 
-    public string AutoCompleteString { get; set; }
+    public string AutoCompleteString => _autoCompleteStringAccessor.CanAcceptCloseBrace ? "}" : null;
```

Comment:
This is the only place where we were using the mutation capabilities of contexts, and it was really just delayed initialization. I've replaced it with an accessor that verifies only one set occurs, and verifies that the set occurred before access.